aptos-move/framework/src/natives/cryptography/algebra/arithmetics/scalar_mul.rs

```diff
@@ -0,0 +1,194 @@
+        let element_handle = safely_pop_arg!($args, u64) as usize;
+        safe_borrow_element!($context, element_handle, $group_typ, element_ptr, element);
+        safe_borrow_element!($context, scalar_handle, $scalar_typ, scalar_ptr, scalar);
+        let scalar_bigint: ark_ff::BigInteger256 = (*scalar).into();
+        $context.charge($gas)?;
+        let new_element = element.$op(scalar_bigint);
+        let new_handle = store_element!($context, new_element);
+        Ok(smallvec![Value::u64(new_handle as u64)])
+    }};
+}
+
+pub fn scalar_mul_internal(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    assert_eq!(2, ty_args.len());
+    let group_opt = structure_from_ty_arg!(context, &ty_args[0]);
+    let scalar_field_opt = structure_from_ty_arg!(context, &ty_args[1]);
+    abort_unless_group_scalar_mul_enabled!(context, group_opt, scalar_field_opt);
+    match (group_opt, scalar_field_opt) {
+        (Some(Structure::BLS12381G1Affine), Some(Structure::BLS12381Fr)) => {
+            ark_scalar_mul_internal!(
+                context,
+                args,
+                ark_bls12_381::G1Projective,
+                ark_bls12_381::Fr,
+                mul_bigint,
+                gas_params.ark_bls12_381_g1_proj_scalar_mul * NumArgs::one()
+            )
+        },
+        (Some(Structure::BLS12381G2Affine), Some(Structure::BLS12381Fr)) => {
+            ark_scalar_mul_internal!(
+                context,
+                args,
+                ark_bls12_381::G2Projective,
+                ark_bls12_381::Fr,
+                mul_bigint,
+                gas_params.ark_bls12_381_g2_proj_scalar_mul * NumArgs::one()
+            )
+        },
+        (Some(Structure::BLS12381Gt), Some(Structure::BLS12381Fr)) => {
+            let scalar_handle = safely_pop_arg!(args, u64) as usize;
+            let element_handle = safely_pop_arg!(args, u64) as usize;
+            safe_borrow_element!(
+                context,
+                element_handle,
+                ark_bls12_381::Fq12,
+                element_ptr,
+                element
+            );
+            safe_borrow_element!(
+                context,
+                scalar_handle,
+                ark_bls12_381::Fr,
+                scalar_ptr,
+                scalar
+            );
+            let scalar_bigint: ark_ff::BigInteger256 = (*scalar).into();
+            context.charge(gas_params.ark_bls12_381_fq12_pow_u256 * NumArgs::one())?;
+            let new_element = element.pow(scalar_bigint);
+            let new_handle = store_element!(context, new_element);
+            Ok(smallvec![Value::u64(new_handle as u64)])
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        }),
+    }
+}
+
+macro_rules! ark_msm_internal {
```

Comment:
Shouldn't this be in `aptos-move/framework/src/natives/cryptography/algebra/arithmetics/mod.rs` with everything else?